aiden-web/aiden_app/services/tools/scraper_tool.py

```diff
@@ -1,29 +1,116 @@
         super().__init__("ScraperTool")
-        self.add_tool("search_jobs", self.search_jobs_wrapper, agent_speaks_next=False)
-        self.profile_embedding_id: UUID = profile_embedding_id
-
-    @Tool.tool_function  # type: ignore
-    def search_jobs_wrapper(self, search_query: str, location: str, num_results: int = 15, *args, **kwargs) -> list[str]:
-        url = f"{os.environ.get('RECOMMENDER_API_URL')}/joboffers"
-        payload = {
-            "location": location,
-            "query": search_query,
-            "limit": num_results,
-            "profile_id": str(self.profile_embedding_id),
-        }
-        response = httpx.post(url=url, json=payload, timeout=60)
+        self.add_tool("search_jobs", self.search_jobs)
+        self.user_profile = qdrant_client.rest.points_api.get_point(collection_name="user_profile", id=str(profile_embedding_id)).result
+        self.returned_offers: dict[UUID, list[str]] = {}
+
+    def _get_search_query_vector(self, search_query: str) -> list[EmbeddingObject]:
+        return mistral_client.embeddings(model="mistral-embed", input=[search_query]).data
+
+    def get_scrape_status(self, scrape_id: UUID) -> ScrapeStatus:
+        response = httpx.get(url=self.scrape_status_url.format(scrape_id=scrape_id), timeout=3)
         response.raise_for_status()
+        return response.json()["status"]
 
-        results = response.json()
+    def get_next_job_recommendation(self, scrape_id: UUID, search_query: str, location: str) -> str:
+        user_profile_vector = self.user_profile.vector
+        search_query_vector = self._get_search_query_vector(search_query + " " + location)[0].embedding
+        search_vector = [a + (b * 0.5) for a, b in zip(search_query_vector, user_profile_vector)]
+
+        filter_condition = {"must_not": [{"key": "reference", "match": {"any": self.returned_offers.get(scrape_id, [])}}]}
+
+        response = qdrant_client.search(
+            collection_name=JOB_COLLECTION,
+            query_vector=search_vector,
+            query_filter=filter_condition,
+            limit=1,
+            with_vectors=False,
+            with_payload=True,
+        )
+
+        if not response:
+            return None  # No more unique job offers available
 
-        return results
+        job_offer = JobOffer(**response[0].payload)
+
+        if scrape_id not in self.returned_offers:
+            self.returned_offers[scrape_id] = []
+        self.returned_offers[scrape_id].append(job_offer.reference)
+
+        return job_offer
+
+    def format_user_message(self, job_offer: JobOffer) -> str:
+        if job_offer.profile is not None:
+            job_offer.profile = markdown2.markdown(job_offer.profile)
+        if job_offer.organization.description is not None:
+            job_offer.organization.description = markdown2.markdown(job_offer.organization.description)
+        return render_to_string("langui/job-offer.html", {"job_offer": job_offer.model_dump()})
```

Comment:
this is the magic behind solving html escapes problem ?

Reply:
Nope, the magic is just sending offers one by one. This is the same parsing that we did before but in a cleaner place.